### ISSUE TITLE
0.6.0 broke Vite dev mode

### ISSUE THREAD

**Issue by gunters63** [jotaijs/jotai-devtools#82]
If i update jotai-devtools to 0.6.0 my UI breaks with the following message in the console:

<img width="545" alt="image" src="https://github.com/jotaijs/jotai-devtools/assets/90346/ddd81387-6877-45ae-9fec-a79f789da50f">

Chalk is an npm package which works only in a Node environment, hence process is not defined.

If I look at the stacktrace, the error comes from this origin:

<img width="952" alt="image" src="https://github.com/jotaijs/jotai-devtools/assets/90346/f8eeaac0-7de0-4b30-be18-2182659c01a7">

Time travel is a new feature in devtools 0.6 if I understand correctly.

Could it be that some of the imported npm packages there use chalk internally? Maybe @mantine/core?

**Comment by gunters63** [comment]
vite-plugin-node-polyfills helped

**Comment by aroman** [comment]
also running into this issue.

by sheer coincidence, last week we needed to build json diff/patching into _our_ project. we ran into issues with jsondiffpatch for exactly this reason, and instead moved to fast-json-patch, which is a much cleaner and more modern API. it's typescript-native. might be worth considering!

**Comment by arjunvegda** [comment]
Thanks for reaching out and for the investigation! It's super helpful

I'm currently away from keyboard for a week, I'll look into this in detail sometime next week.


Meanwhile, is this happening only with Vite? Or are we seeing issues with Rollup and other bundlers too?

**Comment by aroman** [comment]
i did try to give it a shot, but for some reason when I use the local version of jotai-devtools, it's missing all the atoms...

any tips on getting jotai to work locally? 

what I did was:

`pnpm install`
`pnpm compile`

and then I referenced it in my project's package.json like:

```
"jotai-devtools": "../jotai-devtools"
```

I should note that version 0.5.3 from npm works fine, and even version 0.6.0 works fine via npm, using the polyfill plugin that gunters63 mentioned.

<img width="521" alt="CleanShot 2023-06-27 at 02 05 30@2x" src="https://github.com/jotaijs/jotai-devtools/assets/531504/788bde3d-d7ea-4c48-b613-0ffde8b82971">

**Comment by arjunvegda** [comment]
Ah sorry, for the lack of a contribution guide 😅. We use storybook to develop this extension. 

Here are steps to get you up and running locally
- `pnpm install`
- `pnpm storybook`
-  Use `DevTools.stories.tsx` as a playground or `Playground.stories.tsx` to test your changes

**Comment by aroman** [comment]
so, i did a bit more digging here. tl;dr i don't think we should use fast-json-patch. 

RFC 6902 does not include the previous values in the delta operations it generates. this doesn't matter for what I need it for in my project, but now that i've dug into how JSON diffing is being used in this project, it's basically the whole point, making the package of little value (unless you want to forego the entire "diff" view feature, which I don't think you do).

Besides that, looking at the code, it's _very_ tightly coupled to the diff format produced by `jsondiffpatch`. there's a lot of elaborate string and object manipulation code, all with `any` types (since `jsondiffpatch` doesn't include types)... so even if we found an alternative json diff library, unless it happens to output in the same non-standard format as `jsondiffpatch`... it would be a big undertaking to rework everything to use the new format.

Given that... I think the best option is simply using `jsondiffpatch` and patching it to avoid the `process` call. Turns out that's a trivial one-line patch, as someone pointed out [here](https://github.com/benjamine/jsondiffpatch/issues/315#issuecomment-1133773820).

I've created a PR #85 which implements this patch using pnpm.

I can confirm that my patch at least prevents vite from totally breaking, but I can't actually test the functionality because I still get that issue where all the atoms are missing when I install via `npm link jotai-devtools`. I _can_ confirm that it does work locally via `pnpm storybook`, and it passes all the tests.

**Comment by arjunvegda** [comment]
Thanks for your investigation and PR! 

I'm not very confident in shipping a patch via `jotai-devtools` 😅  given that our users have multiple workarounds (define `process.platform` on their end, patch `jsondiffpatch` on userland, etc.). Besides, based on your patch I wonder if it would work for non-pnpm package managers 🤔 

Perhaps, you may be interested in creating a PR in the jsondiffpatch repo for [this issue](https://github.com/benjamine/jsondiffpatch/issues/315) 😄 

> I can't actually test the functionality because I still get that issue where all the atoms are missing when I install via npm link jotai-devtools.

We have a canary deployment setup for PRs, [here are the instructions](https://github.com/jotaijs/jotai-devtools/pull/85#issuecomment-1610878020) on how to install the canary version generated using your PR.

**Comment by aroman** [comment]
agreed with you! the pnpm patching system was new to me too — i tried the canary package and it doesn't work, so looks like the pnpm patch isn't coming through.

I do feel that this is something that should be fixed in jotai-devtools, or at the very least documented, since ultimately jotai-devtools is bundling a dependency with a show-stopping bug in it for vite users.

i agree that the ideal outcome is jsondiffpatch fixes this, but it seems that package has been abandoned since 2021, so I think a PR will fall on deaf ears :(

**Comment by arjunvegda** [comment]
Yes, ideally it'd be nice to address it in `jotai-devtools`. Perhaps we could use [`jsondiffpatch-rc`](https://github.com/benjamine/jsondiffpatch/issues/315#issuecomment-1261719830) until a long-term fix is identified. 🤔 I'll try to brainstorm other ideas.

> but it seems that package has been abandoned since 2021

The author of the `jsondiffpatch` library is [open to having a volunteer](https://github.com/benjamine/jsondiffpatch/issues/274#issuecomment-1579698925). Perhaps there is a beam of hope 😅 .

Chalk dropped the `process` check in `chalk@>=3.0.0`, so maybe the PR would be slightly less complex. WDYT?

**Comment by arjunvegda** [comment]
> Perhaps we could use https://github.com/benjamine/jsondiffpatch/issues/315#issuecomment-1261719830 until a long-term fix is identified. 🤔 I'll try to brainstorm other ideas.

I tried this! Unfortunately, the build fails due to some import error.

---
I [created a PR](https://github.com/benjamine/jsondiffpatch/pull/332) that _should_ fix this on the jsondiffpatch repo 🤞

**Comment by arjunvegda** [comment]
Hey @maylortaylor, just to clarify further, this issue does not completely block you from using DevTools. DevTools does not rely on the code that's causing this error 😁

Here is a workaround you may use in your project. Define `process.platform` in your `vite.config.ts` file until [the PR with the fix](https://github.com/benjamine/jsondiffpatch/pull/332) is merged.

```ts
export default defineConfig({
  // ...
  define: {
    'process.platform': '' // you may set this to anything
  }
})
```

**Comment by arjunvegda** [comment]
Great suggestions! @maylortaylor. Let's continue this discussion in https://github.com/jotaijs/jotai-devtools/discussions/86

**Comment by jokereven** [comment]
> .

I choose to use version 0.5.3

**Comment by Xenfo** [comment]
> Hey @maylortaylor, just to clarify further, this issue does not completely block you from using DevTools. DevTools does not rely on the code that's causing this error 😁
> 
> Here is a workaround you may use in your project. Define `process.platform` in your `vite.config.ts` file until [the PR with the fix](https://github.com/benjamine/jsondiffpatch/pull/332) is merged.
> 
> ```ts
> export default defineConfig({
>   // ...
>   define: {
>     'process.platform': '' // you may set this to anything
>   }
> })
> ```
Just an fyi, this doesn't work with Vitest because then Vitest can't set `process.env` while testing. On top of that `vite-plugin-node-polyfills` causes a lot of terminal output while transforming. So the only way is patching for me atleast.

**Comment by arjunvegda** [comment]
@Xenfo Thanks for sharing your solution! 🙏  

Out of curiosity, why do we need Jotai DevTools when running tests? 🤔

**Comment by Xenfo** [comment]
We don't but the `define` block in the Vite config overrides `process.platform` for all environments, meaning it'll override it when testing with Vitest.

**Comment by arjunvegda** [comment]
Wouldn't it be safer to wrap `<DevTools/>` with a `process.env.NODE_ENV !== 'development'` check then? 🤔

**Comment by aroman** [comment]
> > Hey @maylortaylor, just to clarify further, this issue does not completely block you from using DevTools. DevTools does not rely on the code that's causing this error 😁
> > Here is a workaround you may use in your project. Define `process.platform` in your `vite.config.ts` file until [the PR with the fix](https://github.com/benjamine/jsondiffpatch/pull/332) is merged.
> > ```ts
> > export default defineConfig({
> >   // ...
> >   define: {
> >     'process.platform': '' // you may set this to anything
> >   }
> > })
> > ```
> 
> Just an fyi, this doesn't work with Vitest because then Vitest can't set `process.env` while testing. On top of that `vite-plugin-node-polyfills` causes a lot of terminal output while transforming. So the only way is patching for me atleast.
> 
> Edit: There's actually an even better way to do this, which is overriding packages. Add the following in your `package.json`. I had to do a clean install after for it to work. NPM:
> 
> ```json
>   "overrides": {
>     "jotai-devtools": {
>       "chalk": "^4.1.2"
>     }
>   }
> ```
> 
> Yarn:
> 
> ```json
>   "resolutions": {
>     "jotai-devtools/**/chalk": "^4.1.2"
>   }
> ```
> 
> PNPM:
> 
> ```json
>   "pnpm": {
>     "overrides": {
>       "jotai-devtools>chalk": "^4.1.2"
>     }
>   }
> ```
> 
> Note: I've only tested with NPM.

this worked a charm! for me, using pnpm, what I had to do was:


```json
"pnpm": {
  "overrides": {
    "jsondiffpatch>chalk": "^4.1.2"
  }
}
```

**Comment by Xenfo** [comment]
> Wouldn't it be safer to wrap `<DevTools/>` with a `process.env.NODE_ENV === 'development'` check then? 🤔

Jotai Devtools isn't part of any test code but the define block applies to all environments (including Vitest).
```
    export default defineConfig({
      // ...
      define: {
        'process.platform': '' // you may set this to anything
      }
    })
```
This define block will make `process.platform` readonly (probably by setting it to a const) and Vitest needs to be able to set `process.platform` to run. So it's not actually an issue with the library, just the workaround.

**Comment by arjunvegda** [comment]
Fix released in `v0.6.2` 🚀
